 specification/proto-requirements.md

```diff
@@ -0,0 +1,93 @@
+# OpenTelemetry Protocol Requirements
+
+This document will drive OpenTelemetry Protocol design and RFC.
+
+## Goals
+
+See the goals of OpenTelemetry Protocol design [here](proto-design-goals.md).
+
+## Vocabulary
+
+There are 2 parties involved in telemetry data exchange. In this document the party that is the source of telemetry data is called the Client, the party that is the destination of telemetry data is called the Server.
+
+Examples of a Client are instrumented applications or sending side of telemetry collectors, examples of Servers are telemetry backends or receiving side of telemetry collectors (so a Collector is typically both a Client and a Server depending on which side you look from).
+
+## Known Issues with Existing Protocols
+
+Our experience with OpenCensus and other protocols has been that many of them have one or more of the following drawbacks:
+
+- High CPU consumption for serialization and especially deserialization of received telemetry data.
+- High and frequent CPU consumption by Garbage Collector.
+- Lack of delivery guarantees for certain protocols (e.g. stream-based gRPC OpenCensus protocol) which makes troubleshooting of telemetry pipelines difficult.
+- Not aware / not cooperating with load balancers resulting in potentially large imbalances in horizontally scaled backends.
+- Support either traces or metrics but not both.
+
+Our goal is to avoid or mitigate these known issues in the new protocol.
+
+## Requirements
+
+The following are OpenTelemetry protocol requirements.
+
+#### Supported Node Types
+
+The protocol must be suitable for use between all of the following node types: instrumented applications, telemetry backends, telemetry agents running as local daemons, stand-alone collector/forwarder services.
+
+#### Supported Data Types
+
+The protocol must support traces and metrics as data types.
+
+#### Reliability of Delivery
+
+The protocol must ensure reliable data delivery and clear visibility when the data cannot be delivered. This should be achieved by sending data acknowledgements from the Server to the Client. 
+
+Note that acknowledgements alone are not sufficient to guarantee that: a) no data will be lost and b) no data will be duplicated. Acknowledgements can help to guarantee a) but not b). Guaranteeing both at the same is difficult. Because it is usually preferable for telemetry data to be duplicated than to lose it, we choose to guarantee that there are no data losses while potentially allowing duplicate data.
```

Comment:
Even for a), it is extremely difficult when we look at the entire pipeline.

For example SDK sends data to agent, agent sends data to Stackdriver. If the agent takes incoming data and store them temporarily in-memory (or even on local disk), data loss could happen when agent got unexpected restart (or the machine which agent is running on could get stolen). If the agent is supposed to return a success to the SDK **only if** the data got delivered to Stackdriver (or in another words, the agent could say "I've done my part, now it is Stackdriver's problem"), the agent **has** to actually deliver it to Stackdriver and receive a success acknowledgement. Eventually this system will become a fully distributed transaction system which uses something like two-phase-commit and becomes super slow and fragile.

Reply:
The protocol is concerned with one pair of client/server only. What happens after the server acknowledges the receipt is not a protocol concern. The server may store the data persistently or may discard it, may forward it, etc. It is not a protocol concern. You are correct that unless the protocol requires end-to-end acknowledgement delivery it cannot guarantee delivery end-to-end. However, it is not a goal for the protocol at all. It can be a design goal for a pipeline engineer. We are not designing a pipeline here, we only design telemetry data delivery between a pair of nodes.

This is identical to how TCP guarantees delivery of data but it is not concerned with what the receiver of said data does with the data afterwards. On the "protocol" abstraction level we are not concerned with that.

Reply:
I see where you're coming from, and that sounds totally reasonable.

My worry is that in monitoring space people might misinterpret "no data loss" into different things (e.g. the end-to-end guarantee), it will be great to clarify it.

Reply:
Makes sense, I will add a clarification.

Reply:
Clarification added.

Reply:
Actually, I need to think a bit more about this. Thanks for raising the issue.